File: modules/angular2/src/core/render/dom/shared_styles_host.ts
import {DOM} from 'angular2/src/core/dom/dom_adapter';
import {Inject, Injectable} from 'angular2/src/core/di';
import {SetWrapper} from 'angular2/src/core/facade/collection';
import {DOCUMENT} from './dom_tokens';

@Injectable()
export class SharedStylesHost {
  /** @internal */
  _styles: string[] = [];
  /** @internal */
  _stylesSet = new Set<string>();

  constructor() {}

  addStyles(styles: string[]) {
    var additions = [];
    styles.forEach(style => {
      if (!SetWrapper.has(this._stylesSet, style)) {
        this._stylesSet.add(style);
        this._styles.push(style);
        additions.push(style);
      }
    });
    this.onStylesAdded(additions);
  }

  onStylesAdded(additions: string[]) {}

  getAllStyles(): string[] { return this._styles; }
}

@Injectable()
export class DomSharedStylesHost extends SharedStylesHost {
  private _hostNodes = new Set<Node>();
  constructor(@Inject(DOCUMENT) doc: any) {
    super();
    this._hostNodes.add(doc.head);
  }
  /** @internal */
  _addStylesToHost(styles: string[], host: Node) {
    for (var i = 0; i < styles.length; i++) {
      var style = styles[i];
      DOM.appendChild(host, DOM.createStyleElement(style));
    }
  }
  addHost(hostNode: Node) {
    this._addStylesToHost(this._styles, hostNode);
    this._hostNodes.add(hostNode);
  }
  removeHost(hostNode: Node) { SetWrapper.delete(this._hostNodes, hostNode); }

  onStylesAdded(additions: string[]) {
    this._hostNodes.forEach((hostNode) => { this._addStylesToHost(additions, hostNode); });
  }
}

File: lib/src/platform/dom/shared_styles_host.dart
library angular2.src.platform.dom.shared_styles_host;

import "package:angular2/src/platform/dom/dom_adapter.dart" show DOM;
import "package:angular2/src/core/di.dart" show Inject, Injectable;
import "package:angular2/src/facade/collection.dart" show SetWrapper;
import "dom_tokens.dart" show DOCUMENT;

@Injectable()
class SharedStylesHost {
  /** @internal */
  List<String> _styles = [];
  /** @internal */
  var _stylesSet = new Set<String>();
  SharedStylesHost() {}
  addStyles(List<String> styles) {
    var additions = [];
    styles.forEach((style) {
      if (!SetWrapper.has(this._stylesSet, style)) {
        this._stylesSet.add(style);
        this._styles.add(style);
        additions.add(style);
      }
    });
    this.onStylesAdded(additions);
  }

  onStylesAdded(List<String> additions) {}
  List<String> getAllStyles() {
    return this._styles;
  }
}

@Injectable()
class DomSharedStylesHost extends SharedStylesHost {
  var _hostNodes = new Set<dynamic>();
  DomSharedStylesHost(@Inject(DOCUMENT) dynamic doc) : super() {
    /* super call moved to initializer */;
    this._hostNodes.add(doc.head);
  }
  /** @internal */
  _addStylesToHost(List<String> styles, dynamic host) {
    for (var i = 0; i < styles.length; i++) {
      var style = styles[i];
      DOM.appendChild(host, DOM.createStyleElement(style));
    }
  }

  addHost(dynamic hostNode) {
    this._addStylesToHost(this._styles, hostNode);
    this._hostNodes.add(hostNode);
  }

  removeHost(dynamic hostNode) {
    SetWrapper.delete(this._hostNodes, hostNode);
  }

  onStylesAdded(List<String> additions) {
    this._hostNodes.forEach((hostNode) {
      this._addStylesToHost(additions, hostNode);
    });
  }
}
